### ISSUE TITLE
[Bug report] Multiprocessing fails on Custom Circuit

### ISSUE THREAD

**Issue by QuantumWitness** [scqubits/scqubits#149]
Hi, it's me again.

### Documentation check
If the bug manifests in unexpected behavior (as opposed to a crash), confirm that you have consulted the 
[API documentation](https://scqubits.readthedocs.io/en/latest/api-doc/apidoc.html)

 - [x] I have checked the API documentation.
 - [x] I could not locate relevant information in the documentation or information is missing. 

### Describe the bug
Upon configuring multiprocessing for scqubits and attempting to run for even simple systems (1 node) there appears to be an error that occurs regarding the attribute `operator_func`. Code prints a bunch of error messages then appears to hang until the kernel is stopped.

I attempted to run `plot_evals_vs_paramvals` and `get_spectrum_vs_paramvals` and saw the same behavior. 

Error doesn't occur for the native `Transmon` class.

### Expected behavior
Expect scqubits to complete calculations without error or hangup.

### To Reproduce
Here's my code:

```python
import os

NUM_THREADS = "1"

os.environ["OMP_NUM_THREADS"] = NUM_THREADS
os.environ["OPENBLAS_NUM_THREADS"] = NUM_THREADS
os.environ["MKL_NUM_THREADS"] = NUM_THREADS
os.environ["VECLIB_MAXIMUM_THREADS"] = NUM_THREADS
os.environ["NUMEXPR_NUM_THREADS"] = NUM_THREADS

import numpy as np
import scqubits as scqub
import SQcircuit as sqcirc
import scipy.constants as const
import matplotlib.pyplot as plt
from copy import deepcopy

scqub.settings.NUM_CPUS = 6

tmon_yaml = """# transmon circuit
branches:
- ["JJ", 0, 1, EJ=12,10]
- ["JJ", 1, 0, EJ,10]
- ["C", 1, 0, EC=0.2]
"""

tmon = scqub.Circuit.from_yaml(tmon_yaml, from_file=False,
                               ext_basis="discretized")

flux_array = np.linspace(-1,1,101)

fig, ax=tmon.plot_evals_vs_paramvals('Φ1', flux_array, 
                             evals_count=5, subtract_ground=True)
```

### OS and version used (please complete the following information):
 - OS: maxOS 12.1
 - scqubits version 3.0.1
 - Python version 3.8

### Additional context
The error stacktrace seems like it has a lot of the same issue, so here's one example, rather than the whole stack trace:

```
Process ForkPoolWorker-102:
Traceback (most recent call last):
  File "/Users/QuantumWitness/opt/anaconda3/envs/QIS/lib/python3.8/site-packages/multiprocess/process.py", line 315, in _bootstrap
    self.run()
  File "/Users/QuantumWitness/opt/anaconda3/envs/QIS/lib/python3.8/site-packages/multiprocess/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/QuantumWitness/opt/anaconda3/envs/QIS/lib/python3.8/site-packages/multiprocess/pool.py", line 114, in worker
    task = get()
  File "/Users/QuantumWitness/opt/anaconda3/envs/QIS/lib/python3.8/site-packages/multiprocess/queues.py", line 361, in get
    return _ForkingPickler.loads(res)
  File "/Users/QuantumWitness/opt/anaconda3/envs/QIS/lib/python3.8/site-packages/dill/_dill.py", line 387, in loads
    return load(file, ignore, **kwds)
  File "/Users/QuantumWitness/opt/anaconda3/envs/QIS/lib/python3.8/site-packages/dill/_dill.py", line 373, in load
    return Unpickler(file, ignore=ignore, **kwds).load()
  File "/Users/QuantumWitness/opt/anaconda3/envs/QIS/lib/python3.8/site-packages/dill/_dill.py", line 646, in load
    obj = StockUnpickler.load(self)
AttributeError: 'Circuit' object has no attribute 'operator_func'
```

**Comment by jkochNU** [comment]
Thanks, @QuantumWitness.
Thoughts:
(1) This shows that our pytests are currently not configured to include multiprocessing tests for the `Circuit` class. Let's change that, @saipavanc.
(2) After the fact, the problem is not surprising. Even `dill` has trouble pickling/unpickling a class that has to add methods and bind them dynamically.  I will look into a solution using `__reduce__` and `__setstate__`. That may take a bit of fiddling and time.

**Comment by jkochNU** [comment]
@QuantumWitness Commit https://github.com/scqubits/scqubits/commit/81818ecaadac6a935f809782837f5fde66aec0a4 might address this issue. As you may know, multiprocessing is notoriously finicky (especially on Windows); we do not yet have complete pytest coverage for multiprocessing with `Circuit` for all platforms yet, but you can try whether the current master solves the issue for you.

**Comment by QuantumWitness** [comment]
@jkochNU looks like the commit you referred to has worked! I cloned the repo in a separate environment and re-ran my script with the multiprocessing code and it claims to be computing in parallel. Many thanks, this has made my life much easier.

**Comment by jkochNU** [comment]
@QuantumWitness Our tradition is to close issues once changes have made it into an actual release, so no worries. If your statement that "it claims to be computing in parallel" indicates that there is none of the expected speedup, please follow up. (The naive calculation of time=serial_time/num_cpus is usually a bit too optimistic, and fiddling with thread settings may be needed.)

**Comment by QuantumWitness** [comment]
@jkochNU 

I didn't mean to give the impression that it wasn't being sped up, so I did go and measure the bits of code I was interested in.
6 core speed up gave me between 4.1x - 4.8x speedup, which is good enough for my purposes. Good to know that I might be able to eke out a little more by messing with the thread settings, though.

**Comment by jkochNU** [comment]
This has been addressed with the release of v3.0.2
